### ISSUE TITLE
[Feature] Const constructors?

### ISSUE THREAD

**Issue by ejmount** [zesterer/chumsky#434]
It would be really handy to be able to build a parser as a static item, and this doesn't seem to be precluded by anything obvious so long as we don't try to do the parsing itself in a static context. 

I'm quite happy to have attempt a PR to do this myself. Would this be a welcome feature, or is there some reason it isn't desired?

**Comment by zesterer** [comment]
Definitely welcome! I believe that some of them are already const, but if there are any that aren't but can be made const, I'd be very happy to accept a PR for that!

**Comment by CraftSpider** [comment]
I've tried this before, and the big roadblocks I run into are:

- The fact that parser types are so big makes naming them difficult
- Can't use any boxed or recursive parsers, due to allocation
- `sync` feature needs to be active
- Not a block, but make sure to store a `dyn for<'a> Parser<'a, ...>`, not a `dyn Parser<'static, ...>`

You can avoid the first two with `Lazy`, but then you don't need const-construction anyways

**Comment by CraftSpider** [comment]
Right, and I forgot, since last time I did this, that you may also run into 'implementation is not general enough', which I haven't figured out how to work around.

**Comment by ejmount** [comment]
Unfortunately, it turns out I did miss something in the OP, which is that while you usually know the concrete types involved, the combinators are nonetheless defined on the `Parser` trait, so can't be called in a const context. They also can't be hoisted out of the trait because they're routed through the default impl, you'd end up redefining them for every concrete combinator class.

I had a look through the free functions, and the only ones I found that could easily be const but aren't are `skip_then_retry_until` and `via_parser` which I don't expect have much use in a const context. `todo` could be const if `Location::caller` were const on stable, which is presumably planned. 

The functions in `text.rs` in principle could be const, but they would have to "manually" build the combinators involved, which seems like a lot of code overhead.

**Comment by zesterer** [comment]
We'll just have to wait on const trait methods then, I guess. Unfortunate.
